File: Emulator/src/main/core/game/component/ComponentDefinition.java
package core.game.component;

import java.sql.SQLException;
import java.util.HashMap;
import java.util.Map;

/**
 * The Component definition.
 */
public final class ComponentDefinition {

    private static final Map<Integer, ComponentDefinition> DEFINITIONS = new HashMap<Integer, ComponentDefinition>();

    private InterfaceType type = InterfaceType.DEFAULT;

    private boolean walkable = false;

    private int tabIndex = -1;

    private ComponentPlugin plugin;

    /**
     * Instantiates a new Component definition.
     */
    public ComponentDefinition() {
        // Empty
    }

    /**
     * Parse component definition.
     *
     * @param type     the type
     * @param walkable the walkable
     * @param tabIndex the tab index
     * @return the component definition
     * @throws SQLException the sql exception
     */
    public ComponentDefinition parse(String type, String walkable, String tabIndex) throws SQLException {
        try {
            this.type = InterfaceType.values()[Integer.parseInt(type)];
            this.walkable = Boolean.parseBoolean(walkable);
            this.tabIndex = Integer.parseInt(tabIndex);
        } catch (NumberFormatException e) {
            throw new SQLException("Error parsing ComponentDefinition values", e);
        }
        return this;
    }

    /**
     * For id component definition.
     *
     * @param componentId the component id
     * @return the component definition
     */
    public static ComponentDefinition forId(int componentId) {
        return DEFINITIONS.computeIfAbsent(componentId, id -> new ComponentDefinition());
    }

    /**
     * Put.
     *
     * @param id     the id
     * @param plugin the plugin
     */
    public static void put(int id, ComponentPlugin plugin) {
        ComponentDefinition.forId(id).setPlugin(plugin);
    }

    /**
     * Gets definitions.
     *
     * @return the definitions
     */
    public static Map<Integer, ComponentDefinition> getDefinitions() {
        return DEFINITIONS;
    }

    /**
     * Gets plugin.
     *
     * @return the plugin
     */
    public ComponentPlugin getPlugin() {
        return plugin;
    }

    /**
     * Sets plugin.
     *
     * @param plugin the plugin
     */
    public void setPlugin(ComponentPlugin plugin) {
        this.plugin = plugin;
    }

    /**
     * Gets window pane id.
     *
     * @param resizable the resizable
     * @return the window pane id
     */
    public int getWindowPaneId(boolean resizable) {
        return resizable ? type.resizablePaneId : type.fixedPaneId;
    }

    /**
     * Gets child id.
     *
     * @param resizable the resizable
     * @return the child id
     */
    public int getChildId(boolean resizable) {
        return resizable ? type.resizableChildId : type.fixedChildId;
    }

    /**
     * Gets type.
     *
     * @return the type
     */
    public InterfaceType getType() {
        return type;
    }

    /**
     * Sets type.
     *
     * @param type the type
     */
    public void setType(InterfaceType type) {
        this.type = type;
    }

    /**
     * Is walkable boolean.
     *
     * @return the boolean
     */
    public boolean isWalkable() {
        return walkable;
    }

    /**
     * Sets walkable.
     *
     * @param walkable the walkable
     */
    public void setWalkable(boolean walkable) {
        this.walkable = walkable;
    }

    /**
     * Gets tab index.
     *
     * @return the tab index
     */
    public int getTabIndex() {
        return tabIndex;
    }

    /**
     * Sets tab index.
     *
     * @param tabIndex the tab index
     */
    public void setTabIndex(int tabIndex) {
        this.tabIndex = tabIndex;
    }

    @Override
    public String toString() {
        return "ComponentDefinition [type=" + type + ", walkable=" + walkable + ", tabIndex=" + tabIndex + ", plugin=" + plugin + "]";
    }
}

File: Emulator/src/main/core/game/component/ComponentDefinition.kt
package core.game.component

import java.sql.SQLException

/**
 * Represents the definition of a UI component, including its type, behavior, and associated plugin.
 */
class ComponentDefinition {
    /**
     * The interface type of this component.
     */
    @JvmField
    var type: InterfaceType = InterfaceType.DEFAULT

    /**
     * Indicates whether this component is walkable.
     */
    var isWalkable: Boolean = false

    /**
     * The tab index of this component.
     */
    @JvmField
    var tabIndex: Int = -1

    /**
     * The plugin associated with this component.
     */
    @JvmField
    var plugin: ComponentPlugin? = null

    /**
     * Parses the component definition from string values.
     *
     * @param type     The type identifier as a string.
     * @param walkable Whether the component is walkable, as a string.
     * @param tabIndex The tab index as a string.
     * @return The parsed {@code ComponentDefinition} instance.
     * @throws SQLException If an error occurs while parsing values.
     */
    @Throws(SQLException::class)
    fun parse(
        type: String,
        walkable: String,
        tabIndex: String,
    ): ComponentDefinition {
        try {
            this.type = InterfaceType.values()[type.toInt()]
            this.isWalkable = walkable.toBoolean()
            this.tabIndex = tabIndex.toInt()
        } catch (e: NumberFormatException) {
            throw SQLException("Error parsing ComponentDefinition values", e)
        }
        return this
    }

    /**
     * Retrieves the window pane ID based on the resizable state.
     *
     * @param resizable Whether the interface is resizable.
     * @return The corresponding window pane ID.
     */
    fun getWindowPaneId(resizable: Boolean): Int = if (resizable) type.resizablePaneId else type.fixedPaneId

    /**
     * Retrieves the child ID based on the resizable state.
     *
     * @param resizable Whether the interface is resizable.
     * @return The corresponding child ID.
     */
    fun getChildId(resizable: Boolean): Int = if (resizable) type.resizableChildId else type.fixedChildId

    override fun toString(): String = "ComponentDefinition [type=$type, walkable=$isWalkable, tabIndex=$tabIndex, plugin=$plugin]"

    companion object {
        private val DEFINITIONS: MutableMap<Int, ComponentDefinition> = HashMap()

        /**
         * Retrieves the component definition for the given component ID.
         * If the definition does not exist, a new one is created and stored.
         *
         * @param componentId The component ID.
         * @return The corresponding {@code ComponentDefinition}.
         */
        @JvmStatic
        fun forId(componentId: Int): ComponentDefinition = DEFINITIONS.computeIfAbsent(componentId) { ComponentDefinition() }

        /**
         * Associates a plugin with a component definition.
         *
         * @param id     The component ID.
         * @param plugin The plugin to associate.
         */
        @JvmStatic
        fun put(
            id: Int,
            plugin: ComponentPlugin?,
        ) {
            forId(id).plugin = plugin
        }

        @JvmStatic
        fun put(
            id: Int,
            definition: ComponentDefinition,
        ) {
            DEFINITIONS[id] = definition
        }

        /**
         * Retrieves all component definitions.
         *
         * @return A map of component definitions.
         */
        val definitions: Map<Int, ComponentDefinition>
            get() = DEFINITIONS
    }
}
